pkg/cloud/aws/actuators/machine/actuator.go

```diff
@@ -167,32 +119,110 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
+
+	1) We have not yet attempted to create any control plane VMs
+	2) We have created our first control plane VM, but it's not ready yet (apiserver still booting)
+	3) We have created at least one control plane VM and it's ready
+
+	To avoid race conditions, we use a single ConfigMap to coordinate.
+
+	1) Absence of ConfigMap: if the current machine is a control plane machine, try to create the ConfigMap with that machine's name in the
+	   data.
+			If creation succeeds, proceed to create the VM for the chosen machine.
+			If creation fails because it already exists, move on to checking if the control plane is ready.
+
+	2) Presence of ConfigMap: the chosen control plane machine should either be booting or ready. Try to reach the apiserver.
+			If we can reach the apiserver, we can proceed with a join (can be a control plane or regular node)
+			If we can't reach the apiserver, we need to keep waiting.
+	 */
+
+
+	// Except for the very first control plane machine, everything will be a join, so let's assume we're joining.
+	isNodeJoin := true
+
+	// TODO make sure this name isn't longer than the max allowed length
+	controlPlaneSyncConfigMapName := fmt.Sprintf("%s-controlplane", cluster.Name)
+
+	controlPlaneConfigMap, err := a.coreClient.ConfigMaps(cluster.Namespace).Get(controlPlaneSyncConfigMapName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		if machine.Spec.Versions.ControlPlane == "" {
+			// This isn't a control plane machine - have to wait
+			return errors.New("machine is not a control plane machine and control plane is not operational yet - need to wait")
+		}
+
+		blockOwnerDeletion := true
+		controlPlaneConfigMap = &apicorev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: cluster.Namespace,
+				Name: controlPlaneSyncConfigMapName,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:	cluster.APIVersion,
+						Kind: cluster.Kind,
+						Name: cluster.Name,
+						UID: cluster.UID,
+						BlockOwnerDeletion: &blockOwnerDeletion,
+					},
+				},
+			},
+			Data: map[string]string{
+				"firstMachine": machine.Name,
+			},
+		}
+
+		_, err = a.coreClient.ConfigMaps(cluster.Namespace).Create(controlPlaneConfigMap)
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "error creating control plane sync config map")
+		}
+
+		// This is the first control plane machine, so it's not a join.
+		isNodeJoin = false
+	} else if err != nil {
+		return errors.Wrap(err, "error getting control plane sync config map")
+	} else {
+		// If this is the first control plane machine, there is a good chance that the initial attempt to create the
+		// machine failed because the AWS network wasn't ready yet. We need to therefore check if this is the first
+		// control plane machine, and if so, set isNodeJoin to false so it can be created correctly.
+
+		if controlPlaneConfigMap.Data["firstMachine"] == machine.Name {
```

Comment:
Do we care if the data stored in this configmap becomes stale? This can happen if we delete/ replace the first controlplane machine. Maybe we need to cleanup the config map on delete.

Reply:
(copying my email replies that didn't get slotted in the right place):

~~I don’t believe it matters. Once we get the first control plane node operational, the contents of the data are no longer important. What does matter is that we can check that the data exist so we know it’s a join and not the initial control plane member.~~

That’s an edge case I hadn’t considered and one that needs to be handled. If the control plane is not ready yet and someone deletes this first machine, we need to make sure to clear out the data so a subsequent machine can be considered the initial member.

Reply:
In Cluster.Reconcile you can always check to see if the ID of the lock holder still exists as a machine, and if not, force release the lock.